tes/tes.proto

```diff
@@ -1,599 +1,241 @@
-  string id = 1;
-
-  // OUTPUT ONLY
-  State state = 2;
-
-  // OPTIONAL
-  string name = 3;
-
-  // OPTIONAL
-  string description = 5;
-
-  // OPTIONAL
-  //
-  // Input files.
-  // Inputs will be downloaded and mounted into the executor container.
-  repeated Input inputs = 6;
-
-  // OPTIONAL
-  //
-  // Output files.
-  // Outputs will be uploaded from the executor container to long-term storage.
-  repeated Output outputs = 7;
-
-  // OPTIONAL
-  //
-  // Request that the task be run with these resources.
-  Resources resources = 8;
-
-  // REQUIRED
-  //
-  // A list of executors to be run, sequentially. Execution stops
-  // on the first error.
-  repeated Executor executors = 9;
-
-  // OPTIONAL
-  //
-  // Volumes are directories which may be used to share data between
-  // Executors. Volumes are initialized as empty directories by the
-  // system when the task starts and are mounted at the same path
-  // in each Executor.
-  //
-  // For example, given a volume defined at "/vol/A",
-  // executor 1 may write a file to "/vol/A/exec1.out.txt", then
-  // executor 2 may read from that file.
-  //
-  // (Essentially, this translates to a `docker run -v` flag where
-  // the container path is the same for each executor).
-  repeated string volumes = 10;
-
-  // OPTIONAL
-  //
-  // A key-value map of arbitrary tags.
-  map<string, string> tags = 11;
-
-  // OUTPUT ONLY
-  //
-  // Task logging information.
-  // Normally, this will contain only one entry, but in the case where
-  // a task fails and is retried, an entry will be appended to this list.
-  repeated TaskLog logs = 12;
-
-  // OUTPUT ONLY, REQUIRED
-  //
-  // Date + time the task was created, in RFC 3339 format.
-  // This is set by the system, not the client.
-  string creation_time = 13;
+
+enum State { 
+	UNKNOWN = 0;
+	QUEUED = 1;
+	INITIALIZING = 2;
+	RUNNING = 3;
+	PAUSED = 4;
+	COMPLETE = 5;
+	EXECUTOR_ERROR = 6;
+	SYSTEM_ERROR = 7;
+	CANCELED = 8;
+	PREEMPTED = 9;
+	CANCELING = 10;
 }
 
-enum FileType {
-  FILE = 0;
-  DIRECTORY = 1;
+enum FileType { 
+	FILE = 0;
+	DIRECTORY = 1;
 }
 
-// Input describes Task input files.
-message Input {
-
-  // OPTIONAL
-  string name = 1;
-
-  // OPTIONAL
-  string description = 2;
-
-  // REQUIRED, unless "content" is set.
-  //
-  // URL in long term storage, for example:
-  // s3://my-object-store/file1
-  // gs://my-bucket/file2
-  // file:///path/to/my/file
-  // /path/to/my/file
-  // etc...
-  string url = 3;
-
-  // REQUIRED
-  //
-  // Path of the file inside the container.
-  // Must be an absolute path.
-  string path = 4;
-
-  // REQUIRED
-  //
-  // Type of the file, FILE or DIRECTORY
-  FileType type = 5;
-
-  // OPTIONAL
-  //
-  // File content literal. 
-  // Implementations should support a minimum of 128 KiB in this field and may define its own maximum.
-  // UTF-8 encoded
-  //
-  // If content is not empty, "url" must be ignored.
-  string content = 6;
+enum view { 
```

Comment:
This enum isn't used in the requests which seems wrong.

I know these changes were auto-generated from the openapi spec, but errors like this and the loss of all of the comments makes me wonder if that is the right approach.

Reply:
Maybe try using https://github.com/nytimes/openapi2proto?

Reply:
Or another option: https://github.com/OpenAPITools/openapi-generator

Reply:
The diagram you put together looks correct. 

> as it appears that the view enum is currently being used in a few places:
>   * Here is an example of the view enum value is being used in the Funnel CLI: [task.go]. (https://github.com/ohsu-comp-bio/funnel/blob/release/0.11.0-rc.1/cmd/task/task.go#L149)
>  * Here is where the client is creating the request to send to the sever with the same view enum value [client.go](https://github.com/ohsu-comp-bio/funnel/blob/release/0.11.0-rc.1/tes/client.go#L61)

The more important reason for using enums in the request definitions is so that the generated server will automatically reject requests with other values with an error like: 

```
$ curl localhost:8000/v1/tasks?view=FOOBAR

{
  "error": "FOOBAR is not a valid tes.TaskView",
  "code": 3,
  "message": "FOOBAR is not a valid tes.TaskView"
}
```

If the `view` is defined as a string (like it is in this PR), the validation is left up to you to implement. Here is what I get on your release branch:

```
$ curl localhost:8000/v1/tasks?view=FOOBAR
{
  "next_page_token":  "",
  "tasks":  [
    {
      "creation_time":  "",
      "description":  "",
      "executors":  [],
      "id":  "cl0tvebsaq988ichb7jg",
      "inputs":  [],
      "logs":  [],
      "name":  "",
      "outputs":  [],
      "resources":  null,
      "state":  "EXECUTOR_ERROR",
      "tags":  {},
      "volumes":  []
    }
  ]
}
```

(you can ignore the task error above, I didn't have docker running on my machine when I did this)

> Though this doesn't address the missing comments in tes.proto — I wonder if the alternatives you listed keep those in the generated output?

Yes, it looks like those tools generate comments in the proto file from the descriptions in the openapi definition.

Reply:
Another note: I noticed that views now work a little differently. 

Before:

```
$ curl localhost:8000/v1/tasks/cl0u503saq98itrktup0?view=MINIMAL
{
  "id": "cl0u503saq98itrktup0",
  "state": "EXECUTOR_ERROR"
}
```

On 0.11.0-rc.1

```
$ curl localhost:8000/v1/tasks/cl0u503saq98itrktup0?view=MINIMAL
{
  "creation_time":  "",
  "description":  "",
  "executors":  [],
  "id":  "cl0u503saq98itrktup0",
  "inputs":  [],
  "logs":  [],
  "name":  "",
  "outputs":  [],
  "resources":  null,
  "state":  "EXECUTOR_ERROR",
  "tags":  {},
  "volumes":  []
}
```

Is this intentional? Why send the extra keys with empty values?

Reply:
Ah, definitely not intentional! Hmm...

# Resolution

Seems to be resolved by changing the `EmitUnpopulated` key to `false`  in `server.go`:

[server/server.go#L90](https://github.com/ohsu-comp-bio/funnel/blob/release/0.11.0-rc.1/server/server.go#L90)
```go
	m := protojson.MarshalOptions{
		Indent:          "  ",
		EmitUnpopulated: true,    <-- `true` results in keys being returned with empty values 
		UseProtoNames:   true,        (resulting in the client outputting extra keys)
	}
```

Will go ahead and update this value. 🫡

# Source of Change in gRPC Behavior

Looks like the behavior was changed in the  `grpc-gateway` package:
> We now emit default values for all fields
> https://grpc-ecosystem.github.io/grpc-gateway/docs/development/grpc-gateway_v2_migration_guide/#we-now-emit-default-values-for-all-fields

Related PR:
> runtime: make default marshaler emit default values
> https://github.com/grpc-ecosystem/grpc-gateway/pull/1377

Related issue:
> Support emitting default values in JSON
> https://github.com/grpc-ecosystem/grpc-gateway/issues/233

# Additional Resources

- [(runtime.JSONPb).UnmarshalOptions on pkg.go.dev](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#MarshalOptions)

Reply:
>  If the view is defined as a string (like it is in this PR), the validation is left up to you to implement.

Ah, thank you for explaining that makes perfect sense — will go ahead and update the view from string to enum!

Reply:
Will also add an error handler in the server to reject requests with invalid `view` values (i.e. values not in the `view` enum):

> ```sh
> $ curl localhost:8000/v1/tasks?view=FOOBAR
> 
> {
>   "error": "FOOBAR is not a valid tes.TaskView",
>   "code": 3,
>   "message": "FOOBAR is not a valid tes.TaskView"
> }
> ```

Reply:
> Will also add an error handler in the server

that isn't really needed if you make the changes to the generated proto.

Reply:
## Potential Emit Defaults Issue

Thanks for your pointers on this @adamstruck! Getting things put together on this and running into a potential issue where we want to emit default/empty values for the BASIC and FULL views, but not for the MINIMAL view (where only 'id' and 'state' are emitted).

I'm not sure if protobuf supports this kind of logic using the same marshaler/unmarshaler [2]. Experimenting with a couple possible resolutions (listed below) but more than happy to implement any suitable changes on this.

## Possible Resolutions

- Custom Marshaler à la Grip ([marshaler.go](https://github.com/bmeg/grip/blob/main/server/marshaler.go))?

- Protobuf update with [openapi-generator](https://github.com/OpenAPITools/openapi-generator)?

- Add [FieldMask](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/FieldMask) to MINIMAL view?

- gRPC Interceptor?
  https://edgehog.blog/a-guide-to-grpc-and-interceptors-265c306d3773

- Revert to protobuf v2? 

- Potential future update to Go's JSON encoding?
  https://github.com/golang/go/discussions/63397#discussioncomment-7201222    

## Additional Resources

[1] Change from gRPC v1 to gRPC v2
https://grpc-ecosystem.github.io/grpc-gateway/docs/development/grpc-gateway_v2_migration_guide/

[2] Change from protobuf v2 to protobuf v3
https://github.com/protocolbuffers/protobuf/issues/1606